### ISSUE TITLE
Set up host site for Back End 

### ISSUE THREAD

**Issue by jondhill333** [chingu-voyages/v28-bears-team-10#29]

**Comment by duncanjbain** [comment]
As we're going with NextJS for the frontend, do we require a separate backend running Express and Mongoose or can we get away with using serverless functions with the frontend hosted on Vercel?

From looking through the [documents](https://nextjs.org/docs/api-routes/introduction) on Next.JS API routes it looks like we can get just use serverless functions for all of our API backend rather than running a separate backend NodeJS server.

@guarmo you're working on the backend so I guess this is your call :smile:

**Comment by guarmo** [comment]
During our first meeting, we had agreed on hosting both front and back-end on Heroku but it seems like we're now doing differently?
In that case, @duncanjbain feel free to take on the issue.

**Comment by jondhill333** [comment]
Yes, sorry about the confusion here guys. At that time we weren't confirmed we were using Next.js, and the discussion was about having seperate hosting for FE and BE , and we talked about it just being a good idea to have one provider if possible. It was discussed that this was for possible with Heroku so yes, we should of gone back over that hosting question when we confirmed we were using Next.js.

As Vercel is the company that built Next.js so hosting on there for the FE makes sense, for the reason Duncan says here, it gives you a load of functionality.

So I think actually using Next.js API routes and hosting on Vercel, will actually make building and deploying a lot easier, so to me it makes sense.

**Comment by guarmo** [comment]
Ok, let's host FE and BE separately then. 
Since I have no previous experience with Next JS I'll go through the docs and get familiar with it, only then I'll be able to express a preference regarding serverless functions or running a separate backend server.

**Comment by guarmo** [comment]
So I've just overviewed Next JS today and I believe that, among others, one of its best features is the possibility to use serverless functions for API calls. The way those are structured is very similar to what you would do if creating a separate server, so I don't see why not opting for it really.

**Comment by duncanjbain** [comment]
> So I've just overviewed Next JS today and I believe that, among others, one of its best features is the possibility to use serverless functions for API calls. The way those are structured is very similar to what you would do if creating a separate server, so I don't see why not opting for it really.

Great. One problem that I've come across with using Vercel is that because our repository is part of the Chingu-Voyages organisation, when using the Vercel dashboard to import the repository for deployment we need to use the 'Pro' paid plan as it's an organisational repository, not a personal one.

One way around this is to fork our repository and deploy from that as it will show as a personal repository, the tricky bit is updating the fork each time a commit is made to the main and develop branches.

I'm currently working on a way to get this to work!

**Comment by duncanjbain** [comment]
It looks like Vercel is also limited to 12 serverless functions only, I could see us exceeding this with the API backend so I'll investigate some other options. Looks like Heroku might be the best option after all.

**Comment by jondhill333** [comment]
Hey, sorry can i ask a question so I understand correctly. So a serverless function for us would just be one indidvidual call to one of the API routes? is that right? Sorry I'm a bit unsure on how exactly these work....

So basically we would have one serverless function for each of the  C R U D functionality for any of our datasets :

so potentially 4 serverless functions for each of:

Dev Profile, 
Dev Log-in/Auth, 
Charity profile, 
Charity log-in/Auth 
Charity Job posting.

So potentially 20 serverless functions required if we went with Vercel

Have I understood this correctly? So this is why we would run out on Vercel?

Thanks!

**Comment by duncanjbain** [comment]
> Hey, sorry can i ask a question so I understand correctly. So a serverless function for us would just be one indidvidual call to one of the API routes? is that right? Sorry I'm a bit unsure on how exactly these work....
> 
> So basically we would have one serverless function for each of the  C R U D functionality for any of our datasets :
> 
> so potentially 4 serverless functions for each of:
> 
> Dev Profile, 
> Dev Log-in/Auth, 
> Charity profile, 
> Charity log-in/Auth 
> Charity Job posting.
> 
> So potentially 20 serverless functions required if we went with Vercel
> 
> Have I understood this correctly? So this is why we would run out on Vercel?
> 
> Thanks!

That's my understanding also. I've got automatic deployment to Vercel working now using GitHub actions, I think it's still worth using Vercel for hosting the front end and then having a separate back end on Heroku running express and mongoose etc.

**Comment by jondhill333** [comment]
OK perfect, thanks! And yes for me thats fine

**Comment by guarmo** [comment]
> Hey, sorry can i ask a question so I understand correctly. So a serverless function for us would just be one indidvidual call to one of the API routes? is that right? Sorry I'm a bit unsure on how exactly these work....
> 
> So basically we would have one serverless function for each of the C R U D functionality for any of our datasets :
> 
> so potentially 4 serverless functions for each of:
> 
> Dev Profile,
> Dev Log-in/Auth,
> Charity profile,
> Charity log-in/Auth
> Charity Job posting.
> 
> So potentially 20 serverless functions required if we went with Vercel
> 
> Have I understood this correctly? So this is why we would run out on Vercel?
> 
> Thanks!

My understanding is that each serverless function handles different HTTP methods conditionally within itself like so:

```
export default function devProfile(req, res) {
  if (req.method === 'GET') {
    // GET dev profile
  } else if (eq.method === 'POST') {
   // POST dev profile
  } else {
  // Handle any other HTTP method
  }
}
```

If that's the case I guess we would be able to create all our app's functionalities with 12 functions or less.

**Comment by duncanjbain** [comment]
> > Hey, sorry can i ask a question so I understand correctly. So a serverless function for us would just be one indidvidual call to one of the API routes? is that right? Sorry I'm a bit unsure on how exactly these work....
> 
> > 
> 
> > So basically we would have one serverless function for each of the C R U D functionality for any of our datasets :
> 
> > 
> 
> > so potentially 4 serverless functions for each of:
> 
> > 
> 
> > Dev Profile,
> 
> > Dev Log-in/Auth,
> 
> > Charity profile,
> 
> > Charity log-in/Auth
> 
> > Charity Job posting.
> 
> > 
> 
> > So potentially 20 serverless functions required if we went with Vercel
> 
> > 
> 
> > Have I understood this correctly? So this is why we would run out on Vercel?
> 
> > 
> 
> > Thanks!
> 
> 
> 
> My understanding is that each serverless function handles different HTTP methods conditionally within itself like so:
> 
> 
> 
> ```
> 
> export default function devProfile(req, res) {
> 
>   if (req.method === 'GET') {
> 
>     // GET dev profile
> 
>   } else if (req.method === 'POST') {
> 
>    // POST dev profile
> 
>   } else {
> 
>   // Handle any other HTTP method
> 
>   }
> 
> }
> 
> ```
> 
> 
> 
> If that's the case I guess we would be able to create all our app's functionalities with 12 functions or less.

That's a good point, you could probably do some more clever  conditional checking in each function to further reduce the amount needed in other cases.

Shall we continue with the serverless functions for now and if we need to we can always move them over to an express server?

**Comment by guarmo** [comment]
That sounds good to me! Migrating over a custom server won't take much work in case things go wrong anyways

**Comment by jondhill333** [comment]
Ah perfect, so the repo is good to go then yeah? We can clone it and start pushing up to that repo with features?

**Comment by duncanjbain** [comment]
> Ah perfect, so the repo is good to go then yeah? We can clone it and start pushing up to that repo with features?

Yeah, it should be, I'm going to spend tonight just confirming everything is working as its should with regards to the automatic deployments to Vercel.

**Comment by jondhill333** [comment]
Ok awesome!
